cloudinit/sources/DataSourceAzure.py

```diff
@@ -660,10 +691,249 @@ def setup(self, is_new_instance):
+            # on should happen fast and avoid unnecessary delays
+            # wherever possible.
+            while True:
+                ifname = netlink.wait_for_nic_attach_event(nl_sock,
+                                                           nics_found)
+                nics_found.append(ifname)
+                LOG.info("Detected nic %s attached.", ifname)
+
+                # Attempt to bring the interface's operating state to
+                # UP in case it is not already.
+                self.wait_for_link_up(ifname)
+
+                # If primary nic is not found, check if this is it. The
+                # platform will attach the primary nic first so we
+                # won't be in primary_nic_found = false state for long.
+                if not primary_nic_found:
+                    LOG.info("Checking if %s is the primary nic",
+                             ifname)
+                    (primary_nic_found, expected_nic_count) = (
+                        self._check_if_nic_is_primary(ifname))
+
+                # Exit criteria: check if we've discovered all nics
+                if (expected_nic_count != -1
+                        and len(nics_found) >= expected_nic_count):
+                    LOG.info("Found all the nics for this VM.")
+                    break
+
+        except AssertionError as error:
+            report_diagnostic_event(error, logger_func=LOG.error)
+
+    @azure_ds_telemetry_reporter
+    def _wait_for_all_nics_ready(self):
+        """Wait for nic(s) to be hot-attached. There may be multiple nics
+           depending on the customer request.
+           But only primary nic would be able to communicate with wireserver
+           and IMDS. So we detect and save the primary nic to be used later.
+        """
+        # Report ready if the marker file is not already present.
+        # The nic of the preprovisioned vm gets hot-detached as soon as
+        # we report ready. So no need to save the dhcp context.
+        self._report_ready_if_needed()
+
+        has_nic_been_detached = bool(
+            os.path.isfile(REPROVISION_NIC_DETACHED_MARKER_FILE))
+
+        try:
+            nl_sock = netlink.create_bound_netlink_socket()
+
+            if not has_nic_been_detached:
+                LOG.info("NIC has not been detached yet.")
+                self._wait_for_nic_detach(nl_sock)
+
+            # If we know that the preprovisioned nic has been detached, and we
+            # still have a fallback nic, then it means the VM must have
+            # rebooted as part of customer assignment, and all the nics have
+            # already been attached by the Azure platform. So there is no need
+            # to wait for nics to be hot-attached.
+            if not self.fallback_interface:
+                self._wait_for_hot_attached_nics(nl_sock)
+            else:
+                LOG.info("Skipping waiting for nic attach because we already"
+                         "have a fallback interface")
```

Comment:
```suggestion
            else:
                report_diagnostic_event(
                    "Skipping waiting for nic attach because we already "
                    "have a fallback interface", logger=LOG.info)
                if report_ready_marker_file_present_before_wait_for_all_nics_ready:
                    report_diagnostic_event(
                        'Report ready marker file already existed before cloud-init '
                        'waited for all nics to be ready. A reboot likely occurred.',
                        logger=LOG.info)
```